src/datetime/mod.rs

```diff
@@ -1536,42 +1538,74 @@ impl str::FromStr for DateTime<Local> {
+                let sec = i64::try_from(dur.as_secs()).map_err(|_| OutOfRange::new())?;
+                let nsec = dur.subsec_nanos();
+                (sec, nsec)
+            }
             Err(e) => {
-                // unlikely but should be handled
+                // `t` is before the Unix epoch. `e.duration()` is how long before the epoch it
+                // is.
                 let dur = e.duration();
-                let (sec, nsec) = (dur.as_secs() as i64, dur.subsec_nanos());
+                let sec: u64 = dur.as_secs();
+                let sec: i64 = sec.try_into().map_err(|_| OutOfRange::new())?;
+                let nsec: u32 = dur.subsec_nanos();
                 if nsec == 0 {
+                    // Overflow safety: `sec` was returned by `dur.as_secs()`, and is guaranteed to
+                    // be non-negative. Negating a non-negative signed integer cannot overflow.
                     (-sec, 0)
                 } else {
-                    (-sec - 1, 1_000_000_000 - nsec)
+                    // Overflow safety: In addition to the above, `-x - 1`, where `x` is a
+                    // non-negative signed integer, also cannot overflow.
+                    let sec: i64 = -sec - 1;
+                    // Overflow safety: `nsec` was returned by `dur.subsec_nanos()`, and is
+                    // guaranteed to be between 0 and 999_999_999 inclusive. Subtracting it from
+                    // 1_000_000_000 is therefore guaranteed not to overflow.
+                    let nsec: u32 = 1_000_000_000 - nsec;
+                    (sec, nsec)
                 }
             }
         };
-        Utc.timestamp(sec, nsec).unwrap()
+        Utc.timestamp(sec, nsec).single().ok_or(OutOfRange::new())
     }
 }
 
 #[cfg(feature = "clock")]
-impl From<SystemTime> for DateTime<Local> {
-    fn from(t: SystemTime) -> DateTime<Local> {
-        DateTime::<Utc>::from(t).with_timezone(&Local)
+impl TryFrom<SystemTime> for DateTime<Local> {
+    type Error = OutOfRange;
+
+    fn try_from(t: SystemTime) -> Result<DateTime<Local>, OutOfRange> {
+        DateTime::<Utc>::try_from(t).map(|t| t.with_timezone(&Local))
     }
 }
 
 #[cfg(feature = "std")]
-impl<Tz: TimeZone> From<DateTime<Tz>> for SystemTime {
-    fn from(dt: DateTime<Tz>) -> SystemTime {
-        let sec = dt.timestamp();
+impl<Tz: TimeZone> TryFrom<DateTime<Tz>> for SystemTime {
+    type Error = OutOfRange;
+
+    fn try_from(dt: DateTime<Tz>) -> Result<SystemTime, OutOfRange> {
+        let sec: i64 = dt.timestamp();
+        let sec_abs: u64 = sec.unsigned_abs();
         let nsec = dt.timestamp_subsec_nanos();
         if sec < 0 {
-            // unlikely but should be handled
-            UNIX_EPOCH - Duration::new(-sec as u64, 0) + Duration::new(0, nsec)
+            // `dt` is before the Unix epoch.
+            let mut t =
+                UNIX_EPOCH.checked_sub(Duration::new(sec_abs, 0)).ok_or(OutOfRange::new())?;
```

Comment:
If we're going to call `OutOfRange::new()`, should probably use `ok_or_else()` instead?

Reply:
Ok, done.

I didn't think it would matter either way, since `OutOfRange::new` doesn't do anything at run time. :man_shrugging: